### ISSUE TITLE
Increase entropy of app passwords

### ISSUE THREAD

**Issue by ScottPeterJohnson** [ScottPeterJohnson/purelymail-issues#24]
(This issue was imported from Gitea)
**lunik1** on November 18, 2019:
Currently app passwords are 16 digits, 0-9. According to howsecureismypassword.net this could be brute forced in ~3 days. Would it be possible to include lowercase, uppercase, and some symbols into the charset? This should make the app passwords effectively impossible to brute-force. Easily confusable characters like 0/O/o 1/l/I could be omitted.

**Comment by ScottPeterJohnson** [comment]
Comment by **Scott** on November 18, 2019:
Without access to our database, there's a server limit of approximately 1 password attempt per 10 seconds per IP, so not a high chance of brute forcing there. (About 3 million days under howsecureismypassword's assumptions, given a gigantic botnet?)

Assuming database access (which would be a security breach obviously, but it's best to be paranoid), I think it'd still probably require a number of years with best available hardware to get access to user messages, but you're actually right that it should probably be improved further. 

Part of that will be strengthening the hash function used, but it'd also be a good precaution to lengthen the app passwords too. I think the 16 digit format is the same as Google uses, but Google probably does not actually encrypt anything based on passwords, so we probably need something stronger.

More digits is probably better than mixing in numbers/letters, just because those are harder to type and don't add as much entropy as you'd think.

**Comment by ScottPeterJohnson** [comment]
Comment by **lunik1** on November 18, 2019:
Google's format is 16 characters ([a-z] as far as I can work out) which would be much harder to crack. I think for any device you are using email on it's probably safe to assume that number and letter entry are easy to type, but uppercase and symbols might be overkill.

It is my understanding that increasing the character set of a password is much more effective at increasing the entropy than increasing the password length, but since these passwords should only be used once there's probably not much harm in increasing the length, either.

Of course, all the assumptions made so far assume there is only one valid password. This may not be the case as there could be multiple app passwords and 2FA backup codes.

**Comment by ScottPeterJohnson** [comment]
Comment by **Scott** on November 18, 2019:
Huh, I guess Google must've changed their format.
Anyway, I've updated the site. All new app passwords should generate as 20 digit alphabetic case-insensitive strings, which should provide log2(26)*20 = 94 bits of entropy, as compared to the old log2(10)*16 = 53 bits. 

"howsecureismypassword" rates that at 16 billion years, which should be long enough to safeguard your email until the sun dies out, even with their generous assumptions. (Old app passwords should continue to work too, since I stored a hash.)
